File: backend/src/main/java/com/vocab/service/ProfileService.java
package com.vocab.service;

import com.vocab.dto.request.UpdateProfileRequest;
import com.vocab.dto.response.ProfileResponse;
import com.vocab.entity.User;
import com.vocab.entity.enums.ProgressStatus;
import com.vocab.exception.ResourceNotFoundException;
import com.vocab.repository.DailyWordSetRepository;
import com.vocab.repository.UserRepository;
import com.vocab.repository.UserWordProgressRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class ProfileService {

    private final UserRepository userRepository;
    private final UserWordProgressRepository progressRepository;
    private final DailyWordSetRepository dailyWordSetRepository;
    private final PasswordEncoder passwordEncoder;

    public ProfileResponse getProfile(Long userId) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        long newCount       = progressRepository.countByUserIdAndStatus(userId, ProgressStatus.NEW);
        long learningCount  = progressRepository.countByUserIdAndStatus(userId, ProgressStatus.LEARNING);
        long reviewCount    = progressRepository.countByUserIdAndStatus(userId, ProgressStatus.REVIEW);
        long masteredCount  = progressRepository.countByUserIdAndStatus(userId, ProgressStatus.MASTERED);
        long totalDaysStudied = dailyWordSetRepository.countByUserIdAndCompletedTrue(userId);

        return ProfileResponse.builder()
            .userId(user.getId())
            .username(user.getUsername())
            .email(user.getEmail())
            .displayName(user.getDisplayName())
            .role(user.getRole())
            .streak(user.getStreak())
            .totalWordsLearned(user.getTotalWordsLearned())
            .createdAt(user.getCreatedAt().toLocalDate().toString())
            .newCount(newCount)
            .learningCount(learningCount)
            .reviewCount(reviewCount)
            .masteredCount(masteredCount)
            .totalDaysStudied(totalDaysStudied)
            .build();
    }

    @Transactional
    public ProfileResponse updateProfile(Long userId, UpdateProfileRequest request) {
        User user = userRepository.findById(userId)
            .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        if (request.getDisplayName() != null && !request.getDisplayName().isBlank()) {
            user.setDisplayName(request.getDisplayName().trim());
        }

        if (request.getNewPassword() != null && !request.getNewPassword().isBlank()) {
            if (request.getCurrentPassword() == null ||
                !passwordEncoder.matches(request.getCurrentPassword(), user.getPassword())) {
                throw new IllegalArgumentException("Mật khẩu hiện tại không đúng");
            }
            user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        }

        userRepository.save(user);
        return getProfile(userId);
    }
}

File: frontend/src/app/core/services/profile.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ProfileData {
  userId: number;
  username: string;
  email: string;
  displayName: string;
  role: string;
  streak: number;
  totalWordsLearned: number;
  createdAt: string;
  newCount: number;
  learningCount: number;
  reviewCount: number;
  masteredCount: number;
  totalDaysStudied: number;
}

export interface UpdateProfileRequest {
  displayName?: string;
  currentPassword?: string;
  newPassword?: string;
}

@Injectable({ providedIn: 'root' })
export class ProfileService {
  constructor(private http: HttpClient) {}

  getProfile(): Observable<ProfileData> {
    return this.http.get<ProfileData>(`${environment.apiUrl}/profile`);
  }

  updateProfile(request: UpdateProfileRequest): Observable<ProfileData> {
    return this.http.put<ProfileData>(`${environment.apiUrl}/profile`, request);
  }
}
